### PR TITLE
Editor: Update post URL component

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Part of: https://github.com/WordPress/gutenberg/issues/60291

This PR updates the post url component that is used for updating the post's slug. Designs are from the linked issue: https://github.com/WordPress/gutenberg/issues/60291#issuecomment-2037258428

## Testing Instructions
1. Test the updated component to update the post's slug in both site and post editors
2. Verify no regression is introduced and everything works as before

## Screenshots or screencast 
Screenshot 2024-04-11 at 1 18 19 PM

### CODE REVIEW

File: packages/editor/src/components/post-url/index.js [hunk 1]
```diff
@@ -6,122 +6,146 @@ import { safeDecodeURIComponent, cleanForSlug } from '@wordpress/url';
...
+							'https://wordpress.org/documentation/article/page-post-settings-sidebar/#permalink'
+						) }
+					>
+						{ __( 'Learn more.' ) }
+					</ExternalLink>
+				</div>
+				<div>
+					{ isEditable && (
+						<InputControl
+							__next40pxDefaultSize
```
Comment (ntsekouras) [inline packages/editor/src/components/post-url/index.js:82]: This is probably the most noticeable difference. I replaced `TextControl` with `InputControl`. This way we also have the prefix/suffix support.

**Comment by jasmussen** [comment]
Here's trunk for reference:

trunk url

Here's the PR:

pr

This is better in most ways, but there are a few small things. We should probably still call it URL. Slug, to most people, means a type of snail.

This is technically already an issue, but would be nice to fix. Instead of "Learn more. ↗", the link should be "Learn more↗." with the ↗ before the period.

The URL below needs a bit of space above it, same space as if it were helptext:
Screenshot 2024-04-10 at 16 00 56

So that's 8px:
8px

The space between the / and the slug part of the URL is a bit big. You can probably set that to zero:

Screenshot 2024-04-10 at 15 58 18

**Comment by ntsekouras** [comment]
>We should probably still call it URL. Slug, to most people, means a type of snail.

😆 😆 

If we call it `URL` should it show the full url in the button that triggers the dialog, like in trunk? Because in the designs we also just show the `slug` 🤔

**Comment by jameskoster** [comment]
> If we call it URL should it show the full url in the button that triggers the dialog

That was one of the motivations for using the term 'Slug'. The full URL that you see on trunk isn't very helpful imo. That said I agree it's a bit technical (or zoological 🐌). Is "Post slug" a compromise, or perhaps "Post path", "Post URL"?

The link beneath the input should have help-text sizing (12px), and (ideally) coloring too (#gray-700). If that causes an issue then perhaps we could try something like:

Screenshot 2024-04-10 at 19 15 29

What do you think?

**Comment by ntsekouras** [comment]
>The link beneath the input should have help-text sizing (12px), and (ideally) coloring too (#gray-700). If that causes an issue then perhaps we could try something like:

I think we can do that with the current design with some css. With the alternative suggested design it might make things a bit more complex since we'll try to add the view page link in the control's label row..

Additionally the `page URL` label might be hard to do with i18n. We have a `item_link` label in post types that defaults to `‘Post Link’ / ‘Page Link’, but might not be the best option. 

>This is technically already an issue, but would be nice to fix. Instead of "Learn more. ↗", the link should be "Learn more↗." with the ↗ before the period.

I think we should see this in a different PR as it's about `ExternalLink` component.

So, we need to decide on the label of the panel row and the button's label that triggers the dialog, so I can update the PR and land this.

**Comment by jasmussen** [comment]
> That was one of the motivations for using the term 'Slug'. The full URL that you see on trunk isn't very helpful imo. That said I agree it's a bit technical (or zoological 🐌). Is "Post slug" a compromise, or perhaps "Post path", "Post URL"?

How about Link?

**Comment by ntsekouras** [comment]
>How about Link?

I think that would work for the panel row to be `Link : ${slug}` and in the dialog to have `Link` as a title.

**Comment by jasmussen** [comment]
It would also match the verbiag of the link dialog.

**Comment by ntsekouras** [comment]
Okay, this seems ready for a final review. Thanks all!

File: packages/editor/src/components/post-url/panel.js [hunk 2]
```diff
@@ -44,18 +46,22 @@ export default function PostURLPanel() {
 }
 
 function PostURLToggle( { isOpen, onClick } ) {
-	const label = usePostURLLabel();
+	const slug = useSelect(
+		( select ) =>
+			safeDecodeURIComponent( select( editorStore ).getEditedPostSlug() ),
```
Comment (jorgefilipecosta) [inline packages/editor/src/components/post-url/panel.js]: Minor non blocker: In a case like this I would probably run safeDecodeURIComponent outside useSelect inside a useMemo. On every store change (e.g: character typed) we will be calling safeDecodeURIComponent even if nothing changed on the getEditedPostSlug. It is not problematic and totally fine in one useSelect call but these things add up with hundreds of useSelect calls. Ideally, I think useSelect should not do computations and return things from the store as is.

**Review by jorgefilipecosta — APPROVED** [review]
The code change looks good and worked well on my tests.
There is something I guess we can improve, in a case where the slug is not part of the permalink e.g: in the default plain links we still say the slug forms the last part of the URL which is not true in that case:
Screenshot 2024-04-11 at 10 46 23

Previously we just showed the link without any message:
Screenshot 2024-04-11 at 10 48 17


I guess we could improve and say a message saying slugs are not being used in the URL and that can be changed on the settings or something similar. Can be done in a follow up.

**Comment by jameskoster** [comment]
"Link" works well I think.

I removed the reference to the 'slug' and 'page' (reads oddly when editing a _post_) in the copy.

To Jorge's point, I do wonder if the copy should be help text on the input, so that it isn't displayed when you cannot customise the link. A couple of other considerations:

* Should the popover be a little wider to accommodate longer slugs?
* Should the input be focussed automatically when you open the popover?
* Positionally should it open to the left of the Inspector as other popovers now do (see color picker, design tools panel, etc.)?

Screenshot 2024-04-11 at 10 53 15

**Comment by ntsekouras** [comment]
>To Jorge's point, I do wonder if the copy should be help text on the input, so that it isn't displayed when you cannot customise the link

We can hide it if it's not editable and preserve the current design above the input.

>Should the input be focussed automatically when you open the popover?

We can try that, yes.

> Positionally should it open to the left of the Inspector as other popovers now do 

We can do that too.

**Review by jameskoster — APPROVED** [review]
Sounds good. I'm happy to open follow-up PRs to further finesse the design.

**Comment by ntsekouras** [comment]
I addressed the last feedback except for the auto focus on input and : 
>Should the popover be a little wider to accommodate longer slugs?

We can refine in follow ups..

**Pull request merged**